icicle/src/vec_ops/vec_ops.cu

```diff
@@ -95,25 +95,64 @@ namespace vec_ops {
     E *d_result, *d_alloc_vec_a, *d_alloc_vec_b;
     E* d_vec_a;
     const E* d_vec_b;
+
+    int is_d_alloc_vec_a_allocated = 0;
     if (!config.is_a_on_device) {
-      CHK_IF_RETURN(cudaMallocAsync(&d_alloc_vec_a, n * sizeof(E), config.ctx.stream));
-      CHK_IF_RETURN(cudaMemcpyAsync(d_alloc_vec_a, vec_a, n * sizeof(E), cudaMemcpyHostToDevice, config.ctx.stream));
-      d_vec_a = d_alloc_vec_a;
+      if (config.is_in_montgomery_form) {
+        CHK_IF_RETURN(cudaMallocAsync(&d_alloc_vec_a, n * sizeof(E), config.ctx.stream));
+        CHK_IF_RETURN(cudaMemcpyAsync(d_alloc_vec_a, vec_a, n * sizeof(E), cudaMemcpyHostToDevice, config.ctx.stream));
+        CHK_IF_RETURN(mont::from_montgomery(d_alloc_vec_a, n * sizeof(E), config.ctx.stream, d_alloc_vec_a));
+        is_d_alloc_vec_a_allocated = 1;
+        d_vec_a = d_alloc_vec_a;
+      } else {
```

Comment:
note that in both cases you allocated and copy. You could do it like https://github.com/ingonyama-zk/icicle/blob/621676bd41659bc101f29dc7ac54f29c7cdaa92b/icicle/src/msm/msm.cu#L524

(1) allocate
(2) copy
(3) if mont --> inplace convert